DDF-1635 Refactored core-catalog-standardframework

bdeining opened the pull request:
@dcruver is hero
@wmcnalli 
@shane-voisard

wmcnalli commented:
:+1: 
<!-- Reviewable comment -K2P32XvhUW6nlIJT6fY:-510466975 -->

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

shaundmorris commented on catalog/core/catalog-core-resourcesizeplugin/src/main/java/ddf/catalog/plugin/resourcesize/metacard/MetacardResourceSizePlugin.java:
```diff
@@ -65,7 +66,7 @@ public QueryResponse process(QueryResponse input)
                     } finally {
                         Thread.currentThread().setContextClassLoader(tccl);
                     }
-                } catch (Exception e) {
```
Should we still catch the generic Exception here after CacheException?

bdeining replied:
It seems that the only exception that can be thrown here is the CacheException.  Before CacheException was removed, it was the only exception being caught.  https://github.com/codice/ddf-catalog/commit/e90db6d5d82820a7dfd9d04b0b305c475fb20e6c

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/ResourceCacheInterface.java:
```diff
@@ -39,8 +40,9 @@
+     * @throws CacheException when no valid resource is found in the cache.
      */
-    Resource getValid(String key, Metacard latestMetacard);
+    Resource getValid(String key, Metacard latestMetacard) throws CacheException;
```
Why do we throw an exception on a cache-miss?  This isn't an exceptional situation, it's the normal course of operation for a cache.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java:
```diff
@@ -36,15 +37,16 @@ public CacheKey(Metacard metacard, ResourceRequest resourceRequest) {
-            throw new IllegalArgumentException("ResourceRequest must not be null.");
+            throw new CacheException("ResourceRequest must not be null.");
         }
 
```
If the client code hasn't initialized this object correctly before calling this method, then they've violated the contract of the class.  There are a few ways to prevent this, but throwing a custom checked exception would be my last choice.  If we're going to throw an exception, it should probably be IllegalStateException.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -289,15 +291,17 @@ public void addPendingCacheEntry(ReliableResource reliableResource) {
         if (key == null) {
-            throw new IllegalArgumentException("Must specify non-null key");
+            throw new CacheException("Must specify non-null key");
         }
```
throwing custom checked exceptions for violations of method contracts is going to make for a lot of custom exceptions and try/catch blocks.  If the client is calling the method correctly (with non-null arguments), then they'll never see these.  I think IllegalArgumentException is appropriate here.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -308,7 +312,7 @@ public Resource getValid(String key, Metacard latestMetacard) {
+                throw new CacheException(
                         "Entry found in cache was out-of-date or otherwise invalid.  Will need to be re-cached.  Entry key: "
                                 + key);
             }
```
This is basically a cache entry invalidation, which seems like an odd thing to throw an exception for.  I would expect us to log the invalidation and return null.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -289,15 +291,17 @@ public void addPendingCacheEntry(ReliableResource reliableResource) {
         if (latestMetacard == null) {
-            throw new IllegalArgumentException("Must specify non-null metacard");
+            throw new CacheException("Must specify non-null metacard");
         }
```
see previous comment.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -321,13 +325,13 @@ public Resource getValid(String key, Metacard latestMetacard) {
             LOGGER.debug("No product found in cache for key = {}", key);
-            throw new IllegalArgumentException("No product found in cache for key = " + key);
+            throw new CacheException("No product found in cache for key = " + key);
         }
```
This is a cache miss, which isn't an exceptional situation, so I would expect us to simply return null.  It smells like we're using Exceptions for normal program flow.

dcruver replied:
!fixed

dcruver commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -366,13 +370,13 @@ public boolean containsValid(String key, Metacard latestMetacard) {
-            throw new IllegalArgumentException(
+            throw new CacheException(
                     "Neither the cachedResource nor the metacard retrieved from the catalog can be null.");
         }
```
Method contract violation.  IAE seems appropriate.

dcruver replied:
!fixed

bdeining commented:
Reviewed 9 of 9 files at r1.
Review status: 0 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java, line 51 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaO-r2-51)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java#L51)):</sup>
I would argue that IllegalArgumentException is more appropriate.  According to the Oracle Documentation, IllegalStateExceptions are thrown when a method has been invoked at an illegal or inappropriate time.  In this class, the ctor assigns the resourceRequest and metacard, thus they must have been given illegal arguments.  https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalStateException.html
<!-- Reviewable comment -K2n7FvYU3lqJf37SXbx:-547972982 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 302 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r2-302)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L302)):</sup>
Done.
<!-- Reviewable comment -K2n7WxQh3miPIJcjJdT:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 305 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r2-305)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L305)):</sup>
Done.
<!-- Reviewable comment -K2n7YmoAiiMuJCkVEZ-:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 318 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r2-318)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L318)):</sup>
Done.
<!-- Reviewable comment -K2n7olBEGQovmnzOoXl:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 335 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r2-335)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L335)):</sup>
Done.
<!-- Reviewable comment -K2n7q4QpTIfwFmijDPF:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 381 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r2-381)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L381)):</sup>
Done.
<!-- Reviewable comment -K2n8W0Ldw7Jw6EYfUPU:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/ResourceCacheInterface.java, line 45 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaN-r2-45)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/ResourceCacheInterface.java#L45)):</sup>
Done.
<!-- Reviewable comment -K2n8_RNnJWOWiltDxYq:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

dcruver commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java, line 51 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaO-r2-51)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java#L51)):</sup>
True, but resourceRequest isn't an argument to generateKey().  As the client, calling a method with no arguments and getting IllegalArgumentException is pretty counter-intuitive.  When I suggested IllegalState, I was assuming there were setters that should have been called before generateKey().  Since the object is immutable, I would move these checks to the ctor so the IAE is thrown when the arguments are actually provided.
<!-- Reviewable comment -K2rghcaUVj3DSxLW0Rj:-302176568 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

bdeining commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java, line 51 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaO-r2-51)** ([raw file](https://github.com/codice/ddf/blob/281287a47e6009a51a87bb006fa8a500359af004/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/CacheKey.java#L51)):</sup>
Done.
<!-- Reviewable comment -K2sN9236VjHMPMKJK9A:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

dcruver commented:
Cloned branch, built and tested.  LGTM.

bdeining commented:
Reviewed 7 of 9 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

stustison commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -268,9 +270,9 @@ public void put(ReliableResource reliableResource) {
     public void removePendingCacheEntry(String cacheKey) {
         if (!pendingCache.remove(cacheKey)) {
-            LOGGER.debug("Did not find pending cache entry with key = {}", cacheKey);
+            LOGGER.info("Did not find pending cache entry with key = {}", cacheKey);
```
Why did we change the log level here? This is going to flood the log under high load and cause performance issues.

pklinef replied:
Yeah, these were debug for that reason.

stustison commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java:
```diff
@@ -308,26 +310,25 @@ public Resource getValid(String key, Metacard latestMetacard) {
-            LOGGER.debug("No product found in cache for key = {}", key);
-            throw new IllegalArgumentException("No product found in cache for key = " + key);
+            LOGGER.info("No product found in cache for key = {}", key);
+            return null;
```
Why did we change this to return null instead of throwing an exception. I'm not the biggest fan of this pattern. Now everywhere you call getValid, you need to check that it is null as opposed to just knowing that it is non-null.

stustison commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java:
```diff
@@ -1488,24 +1488,23 @@ protected ResourceResponse getResource(ResourceRequest resourceRequest, boolean
-                LOGGER.error("resource not found", e1);
-                throw new ResourceNotFoundException(e1);
+            String key = new CacheKey(metacard, resourceRequest).generateKey();
+            if (key == null) {
```
As far as I can tell, this will never be true and we've now lost this "Resource not found" exception.

stustison commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/resource/download/ReliableResourceDownloader.java:
```diff
@@ -162,14 +162,12 @@ public ResourceResponse setupDownload(Metacard metacard,
-            } catch (Exception e) {
+            String key = keyMaker.generateKey();
+
+            if (key == null) {
```
This will never be null.

stustison commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java:
```diff
@@ -1488,24 +1488,23 @@ protected ResourceResponse getResource(ResourceRequest resourceRequest, boolean
             if (productCache != null && productCache.containsValid(key, metacard)) {
-                try {
-                    Resource resource = productCache.getValid(key, metacard);
+                Resource resource = productCache.getValid(key, metacard);
```
In the MetacardResourceSizePlugin we changed the TCCL, do we need to do the same thing here? If not, why are we doing it in the plugin?

dcruver commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 331 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r4-331)** ([raw file](https://github.com/codice/ddf/blob/687f4f0773db8f3d3eac828a2056ce0713ab62be/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L331)):</sup>
@Ben - Scott and I discussed this offline and he's ok with this approach.
<!-- Reviewable comment -K2wNkpVrJfJvZilYmqT:2067828558 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

bdeining commented:
Review status: 4 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java, line 273 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaP-r4-273)** ([raw file](https://github.com/codice/ddf/blob/687f4f0773db8f3d3eac828a2056ce0713ab62be/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/impl/ResourceCache.java#L273)):</sup>
Done.
<!-- Reviewable comment -K2wY2lwKdtp4-g1fmuS:-499146009 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java, line 1492 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaQ-r4-1492)** ([raw file](https://github.com/codice/ddf/blob/687f4f0773db8f3d3eac828a2056ce0713ab62be/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java#L1492)):</sup>
Fixed.
<!-- Reviewable comment -K2wY8We0PbYhnJM5lY0:1758359711 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java, line 1498 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaQ-r4-1498)** ([raw file](https://github.com/codice/ddf/blob/687f4f0773db8f3d3eac828a2056ce0713ab62be/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/CatalogFrameworkImpl.java#L1498)):</sup>
TCCL changes no longer exist in the plugin.
<!-- Reviewable comment -K2wY9Wu3l3qNWAOh9Qh:-1977013486 -->

---

<sup>**[catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/resource/download/ReliableResourceDownloader.java, line 167 \[r4\]](https://reviewable.io:443/reviews/codice/ddf/319#-K2OhX-OAfA11qsNNfaR-r4-167)** ([raw file](https://github.com/codice/ddf/blob/687f4f0773db8f3d3eac828a2056ce0713ab62be/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/resource/download/ReliableResourceDownloader.java#L167)):</sup>
Fixed.
<!-- Reviewable comment -K2wYFRdru7Aia8RJKHh:1758359711 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

shane-voisard commented:
Reviewed 3 of 9 files at r3, 1 of 1 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

bdeining commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)
<!-- Sent from Reviewable.io -->

millerw8 commented:
Reviewed 3 of 9 files at r3, 1 of 1 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/319)*
<!-- Sent from Reviewable.io -->
